### ISSUE TITLE
Inconsistent Perplexity Computation

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [pytorch/fairseq#182]
In the train file, get_perplexity is computed by doing 2^nll. In eval_lm, perplexity is computed with e^nll. Assuming the nll in the two files are computed the same way, then one needs to fixed.

**Comment by alexeib** [comment]
nll loss is not computed the same in the two files
in the case of eval_lm, it is the average of scores which are log likelihoods with base e, in the other case it is base 2 (see for example adaptive_loss.py for example). so you will see that the losses are different between eval_lm and what you see during training, but the perplexities are the same

**Comment by hmc-cs-mdrissi** [comment]
I'll try comparing the two perplexities later tomorrow after a model finishes training. The code for adaptive_loss doesn't make that clear to me mainly because it uses adaptive_softmax and that uses log_softmax. log_softmax uses a base e log.

**Comment by alexeib** [comment]
the base is being changed here:
https://github.com/pytorch/fairseq/blob/master/fairseq/criterions/adaptive_loss.py#L68

**Comment by hmc-cs-mdrissi** [comment]
Thanks for the clarification. The perplexities do match, so I'll go ahead and close this.
